### PR TITLE
Python SDR.flatten() fix

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
Previously this used the Reshape class to avoid copying the data.
Now this copies the data into a new object.

The problem was that python would garbage collect the data source
because the Reshape class does not use smart pointers to keep it
alive.  With a data copy into a new object this is no longer an
issue.

I'm begining to think that the Reshape class isn't well designed.
I probably should have listened to my reviewers...

### CODE REVIEW

File: bindings/py/cpp_src/bindings/sdr/py_SDR.cpp [hunk 1]
```diff
@@ -325,8 +325,11 @@ Argument dimensions A list of dimension sizes, defining the shape of the SDR.)",
 R"(See class nupic.bindings.sdr.Reshape)");
 
         py_SDR.def("flatten", [](SDR &self)
-            { return new Reshape(self, {self.size}); },
-R"(See class nupic.bindings.sdr.Reshape)");
+            {
+                auto flat = new SDR({ self.size });
+                flat->setSparse( self.getSparse() );
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/sdr/py_SDR.cpp]: is there c++ flatten to call here?

Comment (ctrl-z-9000-times) [reply bindings/py/cpp_src/bindings/sdr/py_SDR.cpp]: Currently, C++ does not have a flatten method, C++ has reshape method only.

**Comment by ctrl-z-9000-times** [comment]
> Reshape is the most hacky class of SDR

I think part of the issue is that I didn't use smart pointers to connect the reshape/proxy to the main class, so about half of that code deals with managing the lifetimes & cleanup of the objects.

Another thing is that I thought it would be impossible to make a `const SDR` because I didn't know about the `mutable` keyword.  Without the protection of `const`, I didn't want to allow reshaping the SDR object directly for fear that the user would corrupt our internal data.  Now that there is a way to protect an SDR, we could probably allow the user to change an SDRs dimensions.  On the other hand, pybind ignores the `const` keyword so there might still be some issues with changing an SDRs shape.

**Comment by ctrl-z-9000-times** [comment]
Thanks for reviewing this Breznak,

I'm going to have to do some thinking about how to deal with the Reshape, and if as you suggest it should be removed, replaced, or in some way simplified.

**Pull request merged**